File: problems/ramsey-r55/compute/extend_check.c
/* Non-extension of each (5,5,42)-graph in r55_42some.g6 and its complement.

   gcc -O3 -std=c11 -o extend_check extend_check.c
   ./extend_check refs/r55_42some.g6
*/

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define N 42
#define DEG_LO 18
#define DEG_HI 24
#define MAXK 8000

static uint64_t nbr[N];
static uint64_t k4[MAXK];
static uint64_t i4[MAXK];
static int nk4, ni4;
static int assign[N]; /* -1,0,1 */
static unsigned long long nodes, conflicts, decisions;
static int sat_found;

static int pop64(uint64_t x) { return __builtin_popcountll(x); }

static void list_k4(uint64_t *adj, uint64_t *out, int *nout) {
    int a, b, c, d;
    *nout = 0;
    for (a = 0; a < N; a++) {
        for (b = a + 1; b < N; b++) {
            if (!((adj[a] >> b) & 1ULL)) continue;
            uint64_t cab = adj[a] & adj[b];
            for (c = b + 1; c < N; c++) {
                if (!((cab >> c) & 1ULL)) continue;
                uint64_t cabc = cab & adj[c];
                for (d = c + 1; d < N; d++) {
                    if ((cabc >> d) & 1ULL) {
                        if (*nout >= MAXK) {
                            fprintf(stderr, "MAXK overflow\n");
                            exit(2);
                        }
                        out[(*nout)++] =
                            (1ULL << a) | (1ULL << b) | (1ULL << c) | (1ULL << d);
                    }
                }
            }
        }
    }
}

static int propagate(void) {
    int changed = 1;
    while (changed) {
        changed = 0;
        int ones = 0, zeros = 0;
        for (int i = 0; i < N; i++) {
            if (assign[i] == 1) ones++;
            else if (assign[i] == 0) zeros++;
        }
        if (ones > DEG_HI) return 0;
        if (zeros > N - DEG_LO) return 0;
        if (ones == DEG_HI) {
            for (int i = 0; i < N; i++)
                if (assign[i] == -1) {
                    assign[i] = 0;
                    changed = 1;
                }
            if (changed) continue;
        }
        if (zeros == N - DEG_LO) {
            for (int i = 0; i < N; i++)
                if (assign[i] == -1) {
                    assign[i] = 1;
                    changed = 1;
                }
            if (changed) continue;
        }
        for (int t = 0; t < nk4; t++) {
            uint64_t m = k4[t];
            int in_s = 0, nundef = 0, last = -1;
            while (m) {
                uint64_t b = m & -m;
                int v = __builtin_ctzll(b);
                if (assign[v] == 1) in_s++;
                else if (assign[v] == -1) {
                    nundef++;
                    last = v;
                }
                m ^= b;
            }
            if (in_s == 4) return 0;
            if (in_s == 3 && nundef == 1) {
                if (assign[last] == 1) return 0;
                if (assign[last] == -1) {
                    assign[last] = 0;
                    changed = 1;
                }
            }
        }
        for (int t = 0; t < ni4; t++) {
            uint64_t m = i4[t];
            int out_s = 0, nundef = 0, last = -1;
            while (m) {
                uint64_t b = m & -m;
                int v = __builtin_ctzll(b);
                if (assign[v] == 0) out_s++;
                else if (assign[v] == -1) {
                    nundef++;
                    last = v;
                }
                m ^= b;
            }
            if (out_s == 4) return 0;
            if (out_s == 3 && nundef == 1) {
                if (assign[last] == 0) return 0;
                if (assign[last] == -1) {
                    assign[last] = 1;
                    changed = 1;
                }
            }
        }
    }
    return 1;
}

static int score(int v) {
    uint64_t bit = 1ULL << v;
    int s = 0;
    for (int t = 0; t < nk4; t++)
        if (k4[t] & bit) s++;
    for (int t = 0; t < ni4; t++)
        if (i4[t] & bit) s++;
    return s;
}

static int rec(void) {
    nodes++;
    int snap[N];
    memcpy(snap, assign, sizeof(snap));
    if (!propagate()) {
        conflicts++;
        memcpy(assign, snap, sizeof(snap));
        return 0;
    }
    int undefs[N], nu = 0;
    for (int i = 0; i < N; i++)
        if (assign[i] == -1) undefs[nu++] = i;
    if (nu == 0) {
        int ones = 0;
        for (int i = 0; i < N; i++)
            if (assign[i] == 1) ones++;
        if (ones >= DEG_LO && ones <= DEG_HI) {
            sat_found = 1;
            return 1;
        }
        conflicts++;
        memcpy(assign, snap, sizeof(snap));
        return 0;
    }
    int best = undefs[0], bests = -1;
    for (int i = 0; i < nu; i++) {
        int sc = score(undefs[i]);
        if (sc > bests) {
            bests = sc;
            best = undefs[i];
        }
    }
    int after_prop[N];
    memcpy(after_prop, assign, sizeof(after_prop));
    for (int val = 1; val >= 0; val--) {
        memcpy(assign, after_prop, sizeof(after_prop));
        decisions++;
        assign[best] = val;
        if (rec()) return 1;
    }
    memcpy(assign, snap, sizeof(snap));
    return 0;
}

static int parse_graph6(const char *s, uint64_t *out) {
    if (!s || !s[0]) return -1;
    int n = (unsigned char)s[0] - 63;
    if (n != N) return -1;
    memset(out, 0, N * sizeof(uint64_t));
    int need = (N * (N - 1) / 2 + 5) / 6;
    int bitpos = 0;
    for (int j = 1; j < N; j++) {
        for (int i = 0; i < j; i++) {
            int byte = bitpos / 6;
            int off = 5 - (bitpos % 6);
            int bit = (((unsigned char)s[1 + byte] - 63) >> off) & 1;
            if (bit) {
                out[i] |= 1ULL << j;
                out[j] |= 1ULL << i;
            }
            bitpos++;
        }
    }
    (void)need;
    return 0;
}

static void complement_of(const uint64_t *in, uint64_t *out) {
    uint64_t full = (1ULL << N) - 1ULL;
    for (int i = 0; i < N; i++) out[i] = (full ^ (1ULL << i)) ^ in[i];
}

static int check_graph(uint64_t *adj, int *out_k4, int *out_i4) {
    uint64_t cadj[N];
    complement_of(adj, cadj);
    memcpy(nbr, adj, sizeof(nbr));
    list_k4(adj, k4, &nk4);
    list_k4(cadj, i4, &ni4);
    *out_k4 = nk4;
    *out_i4 = ni4;
    for (int i = 0; i < N; i++) assign[i] = -1;
    nodes = conflicts = decisions = 0;
    sat_found = 0;
    rec();
    return sat_found;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "refs/r55_42some.g6";
    FILE *f = fopen(path, "r");
    if (!f) {
        perror(path);
        return 2;
    }
    char line[512];
    int idx = 0;
    int any = 0;
    unsigned long long tot_nodes = 0;
    clock_t t0 = clock();
    printf("idx side k4 i4 sat decisions conflicts nodes sec\n");
    while (fgets(line, sizeof(line), f)) {
        size_t L = strlen(line);
        while (L && (line[L - 1] == '\n' || line[L - 1] == '\r')) line[--L] = 0;
        if (!L) continue;
        uint64_t adj[N], cadj[N];
        if (parse_graph6(line, adj) != 0) {
            fprintf(stderr, "bad graph6 line %d\n", idx);
            return 2;
        }
        complement_of(adj, cadj);
        for (int side = 0; side < 2; side++) {
            uint64_t *g = side ? cadj : adj;
            clock_t t1 = clock();
            int k4c, i4c;
            int sat = check_graph(g, &k4c, &i4c);
            double sec = (double)(clock() - t1) / CLOCKS_PER_SEC;
            tot_nodes += nodes;
            if (sat) any = 1;
            printf("%d %s %d %d %d %llu %llu %llu %.4f\n", idx,
                   side ? "comp" : "stored", k4c, i4c, sat, decisions, conflicts,
                   nodes, sec);
            fflush(stdout);
        }
        idx++;
    }
    fclose(f);
    double sec = (double)(clock() - t0) / CLOCKS_PER_SEC;
    printf("DONE graphs_pairs=%d any_extension=%d tot_nodes=%llu sec=%.3f\n",
           idx, any, tot_nodes, sec);
    return any ? 1 : 0;
}

File: problems/ramsey-r55/compute/extend_check.py
#!/usr/bin/env python3
"""Show that none of the 656 known (5,5,42)-graphs extend by one vertex.

A new vertex v with neighbourhood S subset V(G) yields a (5,5,43)-graph iff
  * 18 <= |S| <= 24
  * S contains no K4 of G          (else v+K4 is a K5)
  * V\\S contains no independent 4-set of G  (else that set plus v is an alpha-5)

This is a 42-variable SAT instance per graph. We solve it by DPLL with the
K4 / independent-4 unit-propagation rules, and write a compact per-graph
trace (decision count, conflict count, runtime). Replay is: run this file.
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

from r55lib import complement, dump_json, list_k_cliques, parse_graph6

ROOT = Path(__file__).resolve().parent
G6_PATH = ROOT / "refs" / "r55_42some.g6"
OUT = ROOT / "certs" / "mckay42_nonextend.json"

# For n=43: deg in [43-25, 24] = [18, 24]
DEG_LO, DEG_HI = 18, 24


class DPLL:
    """3-state SAT on bits 0..n-1. 0=false, 1=true, -1=undef."""

    def __init__(self, n: int, forbid_all: list[int], hit_any: list[int], lo: int, hi: int):
        self.n = n
        self.forbid_all = forbid_all  # bitmasks that cannot be contained in S
        self.hit_any = hit_any  # bitmasks that must meet S
        self.lo = lo
        self.hi = hi
        self.assign = [-1] * n
        self.decisions = 0
        self.conflicts = 0
        self.nodes = 0
        self.sat = False
        self.model = None

    def _propagate(self) -> bool:
        """Unit-propagate. Return False on conflict."""
        changed = True
        while changed:
            changed = False
            ones = 0
            zeros = 0
            undefs = []
            for i, a in enumerate(self.assign):
                if a == 1:
                    ones += 1
                elif a == 0:
                    zeros += 1
                else:
                    undefs.append(i)
            if ones > self.hi:
                return False
            if zeros > self.n - self.lo:
                return False
            # cardinality units
            if ones == self.hi:
                for i in undefs:
                    self.assign[i] = 0
                    changed = True
                if changed:
                    continue
            if zeros == self.n - self.lo:
                for i in undefs:
                    self.assign[i] = 1
                    changed = True
                if changed:
                    continue
            # K4: if 3 verts already in S, 4th must be out
            for mask in self.forbid_all:
                in_s = 0
                last_undef = -1
                n_undef = 0
                m = mask
                conflict = False
                while m:
                    b = m & -m
                    v = b.bit_length() - 1
                    a = self.assign[v]
                    if a == 1:
                        in_s += 1
                    elif a == -1:
                        n_undef += 1
                        last_undef = v
                    m ^= b
                if in_s == 4:
                    return False
                if in_s == 3 and n_undef == 1:
                    if self.assign[last_undef] == 1:
                        return False
                    if self.assign[last_undef] == -1:
                        self.assign[last_undef] = 0
                        changed = True
            # independent 4: if 3 verts already out, 4th must be in
            for mask in self.hit_any:
                out_s = 0
                last_undef = -1
                n_undef = 0
                m = mask
                while m:
                    b = m & -m
                    v = b.bit_length() - 1
                    a = self.assign[v]
                    if a == 0:
                        out_s += 1
                    elif a == -1:
                        n_undef += 1
                        last_undef = v
                    m ^= b
                if out_s == 4:
                    return False
                if out_s == 3 and n_undef == 1:
                    if self.assign[last_undef] == 0:
                        return False
                    if self.assign[last_undef] == -1:
                        self.assign[last_undef] = 1
                        changed = True
        return True

    def solve(self) -> bool:
        def rec() -> bool:
            self.nodes += 1
            if not self._propagate():
                self.conflicts += 1
                return False
            undefs = [i for i, a in enumerate(self.assign) if a == -1]
            if not undefs:
                ones = sum(1 for a in self.assign if a == 1)
                if self.lo <= ones <= self.hi:
                    self.sat = True
                    self.model = list(self.assign)
                    return True
                self.conflicts += 1
                return False
            # branch on a vertex appearing in the most unfinished constraints
            v = max(undefs, key=lambda i: self._score(i))
            for val in (1, 0):
                snap = list(self.assign)
                self.decisions += 1
                self.assign[v] = val
                if rec():
                    return True
                self.assign = snap
            return False

        return rec()

    def _score(self, v: int) -> int:
        bit = 1 << v
        s = 0
        for mask in self.forbid_all:
            if mask & bit:
                s += 1
        for mask in self.hit_any:
            if mask & bit:
                s += 1
        return s


def check_one(nbr: list[int]) -> dict:
    t0 = time.time()
    k4s = list_k_cliques(nbr, 4)
    i4s = list_k_cliques(complement(nbr), 4)
    solver = DPLL(len(nbr), k4s, i4s, DEG_LO, DEG_HI)
    sat = solver.solve()
    return {
        "n_k4": len(k4s),
        "n_ind4": len(i4s),
        "sat": sat,
        "decisions": solver.decisions,
        "conflicts": solver.conflicts,
        "nodes": solver.nodes,
        "seconds": round(time.time() - t0, 4),
        "model_ones": None if not sat else solver.model.count(1),
    }


def main() -> int:
    lines = [ln.strip() for ln in G6_PATH.read_text().splitlines() if ln.strip()]
    results = []
    any_sat = False
    t0 = time.time()
    # Check each stored graph and its complement (656 graphs).
    for side, mapper in (("stored", lambda n: n), ("complement", complement)):
        for i, line in enumerate(lines):
            n, nbr = parse_graph6(line)
            nbr = mapper(nbr)
            rec = check_one(nbr)
            rec["i"] = i
            rec["side"] = side
            results.append(rec)
            if rec["sat"]:
                any_sat = True
                print("SAT", rec, flush=True)
            if (len(results) % 20) == 0:
                print(
                    f"progress {len(results)}/656 last_k4={rec['n_k4']} "
                    f"ind4={rec['n_ind4']} dec={rec['decisions']} "
                    f"sec={rec['seconds']}",
                    flush=True,
                )
    summary = {
        "n_graphs": len(results),
        "any_extension": any_sat,
        "total_seconds": round(time.time() - t0, 3),
        "k4_range": [min(r["n_k4"] for r in results), max(r["n_k4"] for r in results)],
        "ind4_range": [min(r["n_ind4"] for r in results), max(r["n_ind4"] for r in results)],
        "results": results,
    }
    dump_json(str(OUT), summary)
    print(
        f"graphs={len(results)} any_extension={any_sat} "
        f"sec={summary['total_seconds']} wrote {OUT}"
    )
    return 1 if any_sat else 0


if __name__ == "__main__":
    sys.exit(main())
